### ISSUE TITLE
Add a lint for `NonZero`::new(0) literal type footgun

### ISSUE THREAD

**Issue by Sorrowfulwinds** [rust-lang/rust-clippy#17702]
### What it does

Detect if any new() constructor for [NonZeroX variant](https://doc.rust-lang.org/std/num/index.html#types) (NonZeroU8/16/32/64/128/size NonZeroI8/16/32/64/128/size) or any [NonZero<T> variant](https://doc.rust-lang.org/beta/std/num/struct.NonZero.html) where T is any of the Signed or Unsigned integer types is set to 0 and throw an error.

```rust
let footgun: NonZeroU16 = NonZeroU16::new(0).unwrap();
```
Will only fail and panic at run time when this code branch is reached.

### Advantage

- Warns you from effectively typing panic! into your code.

### Example

```rust
let footgun: NonZeroU16 = NonZeroU16::new(0).unwrap();
```
(The key component is any 'NonZeroX::new(0)' should trigger this, the rest is irrelevant to the lint.)

Could be written as:

```rust
panic!("Crash my program please.");
//NonZero type set to zero literal.
```

### Additional Context

Currently NonZero types are unergonomic, this should reasonably be something built into the language itself to allow you to declare foo: NonZeroU16 = 5; at compile time without needing to make a new() call at all but it is necessary for now.
At all levels of optimization except none the compiler does optimize the .unwrap() away to its logical result.

**Comment by ada4a** [comment]
Imo this would be better off as an extension of rustc's built-in [`unconditional_panic`](https://doc.rust-lang.org/rustc/lints/listing/deny-by-default.html#unconditional-panic) lint

**Comment by y21** [comment]
> Imo this would be better off as an extension of rustc's built-in [`unconditional_panic`](https://doc.rust-lang.org/rustc/lints/listing/deny-by-default.html#unconditional-panic) lint

One potential counter-argument: if it were part of `unconditional_panic`, then it would need to actually look at how the result is used and restrict itself to specific panicking-patterns on the result (like immediately `.unwrap()`ing).

A dedicated lint that looks for any `NonZero::new(0)` would be both simpler and more useful imo; it could warn regardless of how the result is used since it's a nonsensical operation either way, e.g. `NonZero::new(0).unwrap_or(NonZero::MIN)` (which `unconditional_panic` couldn't lint, as weird as that example is).

**Comment by ada4a** [comment]
Yeah, that's a valid counterpoint. You're probably right
